### PR TITLE
#6 IO Devices Support

### PR DESCRIPTION
**Pull request by yenoiwesa**

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (yenoiwesa) [inline .vscode/launch.json:1]: I'll remove this vscode specific file.

File: .vscode/tasks.json [hunk 2]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (yenoiwesa) [inline .vscode/tasks.json:1]: Same for this file, will remove.

File: package.json [hunk 3]
```diff
@@ -1,42 +1,47 @@
...
-    }
+  "name": "homebridge-connexoon",
+  "version": "1.3.0",
+  "description": "A Homebridge plugin providing support for the Connexoon, TaHoma and Cozytouch platforms and accessories.",
+  "main": "index.js",
+  "scripts": {
+    "start": "homebridge -D -U ./homebridge-home -P .",
+    "debug": "ndb npm start",
+    "lint": "eslint .",
+    "test": "./node_modules/jest/bin/jest.js"
```
Comment (yenoiwesa) [inline package.json:10]: just doing this should work?
```suggestion
    "test": "jest"
```

File: src/api/device/awning.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        super(json, overkiz)
+        this.isRevert = true;
+    }
+
+    async setPosition(value) {
+        await this.executeCommand(CLOSURE, [this.convertPosition(value)]);
+    }
+
+    getPosition() {
+        let position = this.currentStates.getStateValue(STATE);
```
Comment (yenoiwesa) [inline src/api/device/awning.js:17]:
```suggestion
        const position = this.currentStates.getStateValue(STATE);
```

File: src/api/device/awning.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+const Device = require("./device");
+
+const CLOSURE = 'setDeployment';
+const STATE = 'core:DeploymentState';
+
+class Awning extends Device {
+    constructor(json, overkiz) {
+        super(json, overkiz)
+        this.isRevert = true;
```
Comment (yenoiwesa) [inline src/api/device/awning.js:9]: What does that do?

File: src/api/device/awning.test.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        expect(target.getPosition()).toBe(100);
+    });
+    
+    test('Awning is cap at 1%', async () => {
+        let mockAPI = jest.mock();
+        let target = new Awning({states: [{name: 'core:DeploymentState', value: 0}]}, mockAPI);
+        expect(target.getPosition()).toBe(0);
+    });
+
+    test('Converting position is a noop for awing', () => {
```
Comment (yenoiwesa) [inline src/api/device/awning.test.js:35]: Is it? It's actually identity not noop.

File: src/api/device/awning.test.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        let mockAPI = jest.mock();
+        let target = new Awning({states: [{name: 'core:DeploymentState', value: 100}]}, mockAPI);
+        expect(target.getPosition()).toBe(100);
+    });
+    
+    test('Awning is cap at 1%', async () => {
+        let mockAPI = jest.mock();
+        let target = new Awning({states: [{name: 'core:DeploymentState', value: 0}]}, mockAPI);
+        expect(target.getPosition()).toBe(0);
+    });
```
Comment (yenoiwesa) [inline src/api/device/awning.test.js:33]: Not sure what these are testing. The boundary values? why does it mention  1 and 99% when the test have 0 and 100?

File: src/api/device/device-states.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
+class DeviceStates {
+
+    constructor(states) {
+        this.states = states || [];
```
Comment (yenoiwesa) [inline src/api/device/device-states.js:4]: From the code bellow, it seems like an object would have been more adapted than an array?
There are O(n) lookups for states when you would have had O(1) with object keys.

File: src/api/device/device-states.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        return this.getStateValue("core:OpenClosedState");
+    }
+
+    getStateValue(name) {
+        let state = this.getState(name);
+        return state ? state.value : undefined;
+    }
+
+    setStateValue(name, value) {
+        let state = this.getState(name);
```
Comment (yenoiwesa) [inline src/api/device/device-states.js:25]: `let`s to be replaced by `const` wherever applicable.

File: src/api/device/device-states.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+        if (state) {
+            state.value = value;
+        }
+    }
+
+    hasState(name) {
+        let state = this.getState(name);
+        return !!state;
```
Comment (yenoiwesa) [inline src/api/device/device-states.js:34]: Code style preference: `Boolean(state)` over `!!state`.

File: src/api/device/device-states.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    mergeStates(newStates) {
+        if (!(newStates instanceof DeviceStates)) {
+            throw new TypeError(`Parameter must be of type Array when it is ${typeof newStates}`);
+        }
+
+        newStates.states.forEach(state => {
+            this.replaceOrAddState(state);
+        });
+        return this;
+    }
```
Comment (yenoiwesa) [inline src/api/device/device-states.js:54]: This code would have been much simpler with an Object for the state storage.

File: src/api/device/device.js [hunk 12]
```diff
@@ -27,6 +33,19 @@ class Device {
         return this.json.widget;
     }
 
+    get supportsIOProtocol() {
+        return !!(this.json.controllableName && this.json.controllableName.startsWith('io:'));
```
Comment (yenoiwesa) [inline src/api/device/device.js:37]: Same here regarding `Boolean(..)` over `!!..`

File: src/api/device/device.js [hunk 13]
```diff
@@ -27,6 +33,19 @@ class Device {
...
+    }
+    
+    hasCommand(command) {
+        if (!this.json.definition || !this.json.definition.commands) {
+            return false;
+        }
+        const list = this.json.definition.commands;
+        return !!(list.find((item) => item.commandName == command));
```
Comment (yenoiwesa) [inline src/api/device/device.js:45]: Same here.

Also `list` is too generic of a variable name.

File: src/api/device/exterior-venetian-blind.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    constructor(json, overkiz) {
+        super(json, overkiz)
+    }
+
+    async setPosition(value) {
+        await this.setPositionAndSlateOrientation(value, this.getSlateOrientation());
+    }
+
+    getPosition() {
+        let position = this.currentStates.getStateValue(POSITION_STATE);
```
Comment (yenoiwesa) [inline src/api/device/exterior-venetian-blind.js:17]: Should be `const`

File: src/api/device/exterior-venetian-blind.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        return this.convertPosition(position);
+    }
+
+    resetPosition(value) {
+        this.currentStates.setStateValue(POSITION_STATE, this.convertPosition(value));
+    }
+
+    convertPosition(value) {
+        return 100 - value;
```
Comment (yenoiwesa) [inline src/api/device/exterior-venetian-blind.js:26]: A comment would be good here. Is the position in IO reversed compared to Homekit?

File: src/api/device/roller-shutter.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    constructor(json, overkiz) {
+        super(json, overkiz)
+    }
+
+    async setPosition(value) {
+        await this.executeCommand(CLOSURE, [this.convertPosition(value)]);
+    }
+
+    getPosition() {
+        let position = this.currentStates.getStateValue(STATE);
```
Comment (yenoiwesa) [inline src/api/device/roller-shutter.js:16]: Should be `const`

File: src/api/events/device-state-changed-event.js [hunk 17]
```diff
@@ -0,0 +1,13 @@
+
+const Event = require('./event');
+const EventType = require('./event-type');
+
+class DeviceStateChangedEvent extends Event {
+    constructor(log, timestamp, deviceURL, deviceStates) {
+        super( log, EventType.DeviceStateChangedEventType, timestamp, deviceURL );
```
Comment (yenoiwesa) [inline src/api/events/device-state-changed-event.js:7]: That does not seem to adhere to the Prettier style definition.
I'll make sure to do a run of Prettier on the entire codebase and enforce it with a GitHub Actions build.

File: src/api/events/event-factory.js [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+                break;
+            default:
+                this.log.debug(`Ignore event ${jsonEvent.name}`);
+                events = [];
+        }
+        return events;
+    }
+
+    createDeviceStateChangedEvent(jsonEvent) {
+        let states = new DeviceStates(jsonEvent.deviceStates || []);
```
Comment (yenoiwesa) [inline src/api/events/event-factory.js:38]: Should be `const`

File: src/api/events/event-factory.js [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        return [new DeviceStateChangedEvent(
+            this.log,
+            jsonEvent.timestamp,
+            jsonEvent.deviceURL,
+            states
+        )];
+    }
+
+    createExecutionRegisteredEvent(jsonEvent) {
+        if (jsonEvent.actions == null || jsonEvent.actions.length == 0) {
```
Comment (yenoiwesa) [inline src/api/events/event-factory.js:48]: We have `lodash` in the repository, the code could be changed to:
```suggestion
        if (isEmpty(jsonEvent.actions)) {
```

File: src/api/events/event-factory.js [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+            return [];
+        }
+
+        return jsonEvent.actions
+            .map((action) => this.createExecutionRegisteredEventFromAction(jsonEvent, action))
+            .reduce((prev, curr) => prev.concat(curr));
+    }
+
+    createExecutionRegisteredEventFromAction(jsonEvent, action) {
+        if (action.commands == null || action.commands.length == 0) {
```
Comment (yenoiwesa) [inline src/api/events/event-factory.js:59]: Same here regarding `isEmpty`

File: src/api/events/event-factory.js [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+            this.log,
+            jsonEvent.timestamp,
+            jsonEvent.execId,
+            action.deviceURL,
+            new Command(command.name, command.parameters)
+        ));
+    }
+
+    createExecutionStateChangedEvent(jsonEvent, deviceURLs) {
+        if (deviceURLs == null || deviceURLs.length == 0) {
```
Comment (yenoiwesa) [inline src/api/events/event-factory.js:74]: Same comment here regarding `isEmpty`

File: src/api/events/event-factory.test.js [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+        };
+
+        let events = eventFactory.createEvents(eventJson, ['id1']);
+        expect(events).not.toBe(null);
+        expect(events[0].type).toBe(EventType.ExecutionStateChangedEventType);
+        expect(events[0].id).toBe('id1');
+        expect(events[0].execId).toBe('exec ID');
+        expect(events[0].timestamp).toBe(1581116121345);
+        expect(events[0].newState).toBe('NOT_TRANSMITTED');
+        expect(events[0].oldState).toBe('INITIALIZED');
```
Comment (yenoiwesa) [inline src/api/events/event-factory.test.js:34]: probably better to check the entire shape of `events` at this point.

File: src/api/events/events-controller-factory.js [hunk 23]
```diff
@@ -0,0 +1,7 @@
+const EventsController = require('./events-controller');
+const EventFactory = require('./event-factory');
+
+module.exports = function(log, overkiz) {
+    let eventFactory = new EventFactory(log);
```
Comment (yenoiwesa) [inline src/api/events/events-controller-factory.js:5]: Should be `const`

File: src/api/events/events-controller.js [hunk 24]
```diff
@@ -0,0 +1,204 @@
+const EventType = require('./event-type');
+
+const MAX_EXPONENT = 7;
+
+class EventsController {
+
+    constructor(log, eventFactory, overkiz, activeFrequencyMs = 5 * 1000, refreshFrequencyMs = 10 * 1000, executionGarbageCollectorCMs = 3 * 60 * 1000) {
```
Comment (yenoiwesa) [inline src/api/events/events-controller.js:7]: There defaults would be better suited in constants

File: src/api/events/events-controller.js [hunk 25]
```diff
@@ -0,0 +1,204 @@
+const EventType = require('./event-type');
```
Comment (yenoiwesa) [inline src/api/events/events-controller.js:1]: I'll need to dig deeper into the functionality of that class.
Does it work with RTS devices too?

File: src/api/events/execution-registered-event.js [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+    get targetPosition() {
+        if (!this.isAffectingPosition) {
+            return 0;
+        }
+
+        return parseInt(this.command.parameters[0].value) || 0;
+    }
+
+    get isAffectingPosition() {
+        let cmd = this.command.name;
```
Comment (yenoiwesa) [inline src/api/events/execution-registered-event.js:22]: Should be `const`

File: src/api/events/execution-registered-event.js [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+
+        return ['setClosureAndOrientation', 'setPosition', 'setClosure', 'setDeployment'].includes(cmd);
```
Comment (yenoiwesa) [inline src/api/events/execution-registered-event.js:24]: Should be a constant (and a `Set`)

File: src/api/overkiz-api.js [hunk 28]
```diff
@@ -20,30 +19,98 @@ class OverkizAPI {
     }
 
     getUrlForQuery(query) {
-        return this.handler.server.getUrlForQuery(query);
+        return this.requestHandler.server.getUrlForQuery(query);
+    }
+
+    async registerEvents() {
+        try {
+            let registration = await this.requestHandler.sendRequestWithLogin(request =>
```
Comment (yenoiwesa) [inline src/api/overkiz-api.js:27]: Should be `const`

File: src/connexoon-platform.js [hunk 29]
```diff
@@ -23,54 +25,49 @@ class ConnexoonPlatform {
...
         } finally {
             callback(this.platformAccessories);
         }
+
+        await this.eventsController.start();
+    }
+
+    _registerDevice(device) {
+        const config = get(this.config, ['devices', device.name]);
+        let service = serviceFactory({homebridge, log: this.log, eventsController: this.eventsController, device, config});
```
Comment (yenoiwesa) [inline src/connexoon-platform.js:56]: Should be `const`

File: src/connexoon-platform.test.js [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+            new Device({ uiClass: "Awning", label: "Awning", deviceURL: "6" }, null),
+        ];
+
+        mockOverkizApi.listDevices = jest.fn().mockResolvedValue(listDevices);
+        connexoonPlatform.accessories(accessories => {
+            expect(accessories).toHaveLength(1);
+            done();
+        });
+    });
+    */
```
Comment (yenoiwesa) [inline src/connexoon-platform.test.js:58]: Commented unit test to be removed

File: src/services/io-window-covering.js [hunk 31]
```diff
@@ -0,0 +1,250 @@
...
+
+        eventsController.subscribe(device.id, this.onEvent.bind(this));
+        this.log(`Creating window covering ${this.name}`);
+    }
+
+    _createHAPService() {
+        return new Service.WindowCovering(this.name);
+    }
+
+    _mixinHAPServiceCharacteristics() {
```
Comment (yenoiwesa) [inline src/services/io-window-covering.js:38]: I am not clear why this method is called `mixin`

File: src/services/io-window-covering.js [hunk 32]
```diff
@@ -0,0 +1,250 @@
...
+        }
+    }
+
+    onDeviceStateChagned(event) {
+        if (!event.deviceStates.hasPositionState) {
+            // only care about position state at the moment
+            return;
+        }
+
+        let prevPosition = this.device.getPosition();
```
Comment (yenoiwesa) [inline src/services/io-window-covering.js:84]: Check `let`s vs `const`s everywhere in this file.

File: package.json (hunk 3 above)
Comment (renajohn) [reply package.json:10]: This is true, but only if jest is installed globally. This is going to break the docker file, though.

File: src/api/device/device-states.js (hunk 8 above)
Comment (renajohn) [reply src/api/device/device-states.js:4]: This is the way the states are returned by the server. Maybe you can get the same state key returned in a given array.

File: src/api/device/exterior-venetian-blind.js (hunk 15 above)
Comment (renajohn) [reply src/api/device/exterior-venetian-blind.js:26]: it is.

File: src/api/events/events-controller.js (hunk 24 above)
Comment (renajohn) [reply src/api/events/events-controller.js:7]: Constants are harder to test. Ideally, you want to have those value injected into your class to mock them.

**Comment by renajohn** [comment]
I did cut corners here and there, so a bit of a refactoring is usually a good thing.
At the moment, my system works for my setup, without restart since my last commit. This is enough for me to put it in the category: "it works, so I won't touch it" :)
Realistically speaking, I won't have the time to do any modifications on the code. If you want to merge, feel free to modify all the code you want.

**Comment by yenoiwesa** [comment]
Closing this pull request based on the comment made on the related issue: https://github.com/yenoiwesa/homebridge-connexoon/issues/6#issuecomment-647338891
